specification/dartLangSpec.tex

```diff
@@ -5193,52 +5195,70 @@ \subsubsection{Correct Member Overrides}
 \begin{itemize}
 \item
   $m$ and $m'$ are both methods, both getters, or both setters.
-  % We cannot have a setter and a method, say, because they cannot have
-  % the same name. However, we must _allow_ $m$ and $m'$ to be two setters.
+\item
+  If $m$ and $m'$ are both getters:
+  The return type of $m$ must be a subtype of the return type of $m'$.
 \item
   If $m$ and $m'$ are both methods or both setters:
   Let $F$ be the function type of $m$
   except that the parameter type is the built-in class \code{Object}
-  for each parameter of $m$ which has the modifier \COVARIANT.
+  for each parameter of $m$ which is covariant-by-declaration
```

Comment:
Do we only do this for covariant-by-declaration and not covariant-by-generics?

(Or handle covariant-by-generics in genereal: What if `m` is the signature of `void add(int x) {}` and `m'` is `add` from `List<int>`? It seems that `m` is still not a valid implementation of `m'` when cast to `List<Object>`.)

What is the real difference between the two?
* Covariant by declaration: Inherited, allows sub-type as parameter type in override.
* Covariant by generics: Inherited, does not allow sub-type as parameter type in override.

That suggests that we do need to distinguish the two in the "valid override" relation, but only in the case where the overriding method's parameter type is a subclass of the overridden one, in all other contexts, they should be treated equivalently?

Not sure. I'll think some more and come back.

Reply:
Yes, with covariant-by-class we require the tools to create forwarding methods as needed, cf. https://github.com/dart-lang/sdk/issues/31580#issuecomment-350420199, we only report an error for parameters that are covariant-by-declaration.

Reply:
> only in the case where the overriding method's parameter type is a subclass of the overridden one,

This would be a global property in any case: If the inherited concrete member has a parameter type which is a supertype of the overridden one, it could still be a non-supertype of an indirectly overridden one.

```dart
class A1 {}
class A2 {}
class B implements A1, A2 {}

class C {
  void foo(A1 b) { print(b.runtimeType); }
}

abstract class D {
  void foo(A2 b);
}

abstract class E implements D {
  void foo(covariant B c);
}

class F extends C implements E {}

void main() {
  D d = F();
  d.foo(A2());
}
```

Reply:
Argh.
So what are the rules that we actually want to specify here? I've lost track.

The *situation* we want to avoid is that a normal implementation method ends up implementing a covariant interface signature. That's not *necessarily* a problem (maybe there are no super-classes to upcast to which would make it unsound), but we see it as *potentially* unsound and want to refuse it. (We currently don't. AFAICS, both VM and dart2js introduce forwarders, at least in the simple cases I tested. Why don't we just specify that and call it a day?)

If that happens, the rules we try to specify here trigger and make it a compile-time error because a non covariant-by-declaration implementation tries to implement a covariant-by-declaration interface, and is presumably not able to handle the potential unsound upcast.
So, the rule should say that:
* if *p'* is covariant-by-declaration and *p* can't accept all arguments (which it can if its type is a top-type, or if it's covariant-by-generics or covariant-by-declaration, because then we know that it's prepared to be used in unsound situations), then *m* is not a valid override of *m'*.

So, something like:

> A member signature *m* is a valid override of a member signature *m'* iff all the following conditions hold:
> 
> * *m* and *m'* are either both getter signatures, both setter signatures or both method signatures.
> * If *m* and *m'* are getter or method signatures, then the return type of *m* is a subtype of the return type of *m'*.
> * If *m* and *m'* are setter or method signatures, then
>   * For each parameter *p* of *m*, if there is no corresponding parameter of *m'*, then *p* is optional.
>   * For each parameter *p'* of *m'* there is a corresponding parameter *p* of *m*, such that:
>     * If *p* is required, so is *p'*,
>     * Either 
>       * the type of *p* is a supertype of the type of *p'*, or
>       * the type of *p* is a subtype of the type of *p'* and *p* is covariant-by-declaration.
>     * If *p'* is covariant-by-declaration, then either *p* is covariant-by-declaration, *p* is covariant-by-generics, or the type of *p* is a top-type.

I wouldn't change the covariant parameter types of `m` to `Object` early. It seems lossy, and you never known when you'd need the information. For example, it won't be able to check the "subtype and covariant-by-declaration" rule here.

Reply:
I made some tests: https://dartpad.dev/?id=91a6cfc411dc4cdf9c392ea49a62b9aa&

It does seem that we currently add forwarders for both covariant-by-generics and covariant-by-declaration.
Should we just accept and specify that? It seems to work, and anything touching `covariant` is prone to run-time errors, so I wouldn't expect a compile-time error.

Reply:
> So what are the rules that we actually want to specify here

We're specifying the relationship between the properties of two corresponding formal parameters in a correct override. One property is the declared type, the other one is covariance-by-declaration.

When the parameter is covariant-by-declaration, the types are checked by the rule in section 'Classes' (so we disable that check here by building an `$F$` with a parameter type `Object`/`Object?` for those parameters). When it isn't, we rely on the function type subtype rules.

> avoid [..] normal implementation [..] implementing a covariant interface signature. That's not necessarily a problem

I agree that we may be able to carve out a little bit of extra flexibility: No dynamic check is required when the overriding parameter has a type which is a supertype of all the types of the (directly or indirectly) overridden parameters (covariant or not, the important point is simply that we must take into account all possible static types of the receiver where that member could be called).

> (We currently don't. AFAICS, both VM and dart2js introduce forwarders,
> at least in the simple cases I tested. Why don't we just specify that and call it a day?)

I argued many times that we should do that, but the decision was to not do it, cf. https://github.com/dart-lang/sdk/issues/31580#issuecomment-350420199. ;-)

But we do get the error in some cases, including Paul's original example:

```
class A {}
class B extends A {}
class C {
  void f(B x, B y) {}
}
abstract class I1 {
  void f(covariant A x, B y);
}
class D extends C implements I1 {}
```

The analyzer flags `D` as an error, indicating that it is not a correct override.

> So, something like:

We could rewrite these rules to re-state the function type subtype requirements, but that introduces a substantial amount of redundancy that we must then maintain, and it may not be obvious to a reader that this is all the same stuff that we're saying about subtypes, repeated.

> `the type of p is a subtype of the type of p' and p is covariant-by-declaration.`

This is redundant with the all-superinterfaces rule, so that's even more redundancy to keep in sync (and I do not think we should change 'correct override' to be a relationship between a member signature and _all_ the member signatures that it overrides, because we don't want to imply that all the checks must be made globally&mdash;the global check should preferably only be performed in the presumably rare cases where we actually have a cov-by-decl parameter).

> `either p is covariant-by-declaration, p is covariant-by-generics`

It is again an interesting generalization that we could allow the inheritance when `p` is cov-by-class, because that method must then already perform a dynamic check on the actual argument type.

Reply:
But we do introduce forwarders in some cases, like when only the covariance marker differs, but the type without the covariance is actually fine. Do we really want to make that an error instead? (It seems reasonable to users for all the same reasons it does for covariance-by-generic).

So, if the function signature of *m* (with `Object` for covariant parameters, either by-declaration or both) is a subtype of the function signature of *m'*, there is no problem, even if *p'* is covariant-by-generics.
That seems to be what we have implemented.

If it's not a subtype, then it must be because a parameter *p* of *m*, has a type that is not a supertype of the type of the corresponding parameter *p'* of *m'*. Since we did the `Object` change, that means *p* is not covariant. So, we catch that case automatically. 

That only leaves the case where the type of *p* is a supertype of the type of *p'*, *p* is not covariant and *p'* is ... which is not currently an error, and which you are trying to make one. Not sure it's worth it.

Reply:
I'd be quite happy about asking for forwarding methods whenever needed, and then we wouldn't have the cov-by-decl error at all.

Reply:
But it is not exactly what we have implemented: Paul's original example is rejected at compile-time.

Reply:
In that example (if it's the one above), the interface requires `A` and the implementation only accepts the subtype `B`. So, that's not a valid implementation of the interface if you ignore covariance, and you ignore covariance because `C`'s implementation method is not coviriant.
If the implementation had accepted an `A` as argument, then it would be sound ignoring covariance, and then the forwarder would just be for allowing upcasts, not changing the visible implementation from accepting `B` to accepting `A`.

Reply:
Actually, the CFE could cut down on the forwarding stubs: The only situation where a cov-by-decl parameter `p` with type `T` must have a dynamic check is when there is a supertype of the receiver where that parameter has a non-subtype of `T`.

For instance, the CFE would not have to generate the forwarder in the following case:

```dart
class A {}

class B extends A {}

class C {
  void f(A x) {}
}

abstract class I1 {
  void f(covariant A x);
}

class D extends C implements I1 {}

void main() {}
```

But it does generate that forwarder:

```
library /*isNonNullableByDefault*/;
import self as self;
import "dart:core" as core;

class A extends core::Object {
  synthetic constructor •() → self::A
    : super core::Object::•()
    ;
}
class B extends self::A {
  synthetic constructor •() → self::B
    : super self::A::•()
    ;
}
class C extends core::Object {
  synthetic constructor •() → self::C
    : super core::Object::•()
    ;
  method f(self::A x) → void {}
}
abstract class I1 extends core::Object {
  synthetic constructor •() → self::I1
    : super core::Object::•()
    ;
  abstract method f(covariant self::A x) → void;
}
class D extends self::C implements self::I1 {
  synthetic constructor •() → self::D
    : super self::C::•()
    ;
  forwarding-stub method f(covariant self::A x) → void
    return super.{self::C::f}(x);
}
static method main() → void {}
```

Reply:
> If the implementation had accepted an A as argument, then it would be sound ignoring covariance

This is a non-local property: We're comparing the type of the formal parameter of the implementation to all the types of the corresponding formal parameters of overridden declarations, directly or indirectly.

So if we want to make a distinction between "good" and "bad" occurrences of inheritance where covariance-by-decl is added to a parameter of an implementation, it can't be expressed in terms of a correct override relation to a direct superinterface, it must be in terms of the complete set of superinterfaces.

We did not make that distinction when we discussed this topic.

I think it would be considerably nicer to allow addition of covariance-by-decl implicitly, and ask tool teams to generate the required forwarders, rather than starting to carve out all these exceptions: The exceptions will just be some cases where an implementation doesn't have to generate a forwarder anyway.

Reply:
The newest commit removes the criterion about covariant parameters and adds a sentence about forwarding methods to the commentary.

It is then up to the implementations to detect exactly where there is a need for a forwarding method, and various carved out pieces of flexibility are now a matter of optimization: Developers are allowed to rely on covariance-by-declaration being added to an inherited method, just like covariance-by-class was always added as needed.

Reply:
@lrhn, I landed this before the vacation, but we did have some ongoing discussions. Do you wish to take another turn in another PR?

Reply:
Let's discuss what we want before writing more spec.

I remember loosing track a few times, back when I could remember what this even was, so let's go slowly and make sure we are talking about the same thing. :)

Reply:
Sounds good! I'll close this thread because the further discussions will occur elsewhere.